recipes/libmad/all/conanfile.py

```diff
@@ -0,0 +1,80 @@
+from conans import ConanFile, tools, AutoToolsBuildEnvironment, MSBuild
+from conans.errors import ConanInvalidConfiguration
+import os
+
+
+class LibmadConan(ConanFile):
+    name = "libmad"
+    description = "MAD is a high-quality MPEG audio decoder.format."
+    topics = ("conan", "mad", "MPEG", "audio", "decoder")
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://www.underbit.com/products/mad/"
+    license = "GPL-2.0-or-later"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {"shared": [True, False], "fPIC": [True, False]}
+    default_options = {"shared": False, "fPIC": True}
+    _source_subfolder = "source_subfolder"
+    _build_subfolder = "build_subfolder"
+
+    def config_options(self):
+        if self.settings.os == 'Windows':
+            del self.options.fPIC
+
+    def configure(self):
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+        if self._is_msvc and self.options.shared:
+            raise ConanInvalidConfiguration("libmad does not support shared library for MSVC")
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    @property
+    def _is_msvc(self):
+        return self.settings.compiler == "Visual Studio"
+
+    def build(self):
+        if self._is_msvc:
+            self._build_msvc()
+        else:
+            self._build_configure()
+
+    def _build_msvc(self):
+        with tools.chdir(os.path.join(self._source_subfolder, "msvc++")):
+            # cl : Command line error D8016: '/ZI' and '/Gy-' command-line options are incompatible
+            tools.replace_in_file("libmad.dsp", "/ZI ", "")
+            if self.settings.arch == "x86_64":
+                tools.replace_in_file("libmad.dsp", "Win32", "x64")
+                tools.replace_in_file("libmad.dsp", "FPM_INTEL", "FPM_DEFAULT")
+                tools.replace_in_file("mad.h", "# define FPM_INTEL", "# define FPM_DEFAULT")
+            with tools.vcvars(self.settings):
+                self.run("devenv libmad.dsp /upgrade")
+            msbuild = MSBuild(self)
+            msbuild.build(project_file="libmad.vcxproj")
+
+    def _build_configure(self):
+        with tools.chdir(self._source_subfolder):
+            if self.options.shared:
+                args = ["--disable-static", "--enable-shared"]
+            else:
+                args = ["--disable-shared", "--enable-static"]
+            env_build = AutoToolsBuildEnvironment(self)
+            env_build.configure(args=args)
+            env_build.make()
+            env_build.install()
+            la = os.path.join(self.package_folder, "lib", "libmad.la")
+            if os.path.isfile(la):
+                os.unlink(la)
```

Comment:
Can't we have these steps in the package() method?

Reply:
we can, but it works both ways, otherwise one more condition in `package`

Reply:
Not saying it is a stopper, but at first sight, I thought the `package()` method was missing the packaging for Linux and realized it was finally here